### PR TITLE
[WIP] Cassandra support via JPA

### PR DESCRIPTION
**Pull request by tduehr**
Adds support for Cassandra via Hibernate OGM for NoSQL databases. Also allows other NoSQL libraries supported by Hibernate OGM to be added by simply adding the correct dependencies.

Waiting release of 5.3.0 for hibernate-ogm-cassandra. Testing currently against a PR version. Must wait for CAS 6.0 due to required schema changes in #3149.

Ensure that you include the following:

- [x] Brief description of changes applied
- [x] Any documentation on how to configure, test
- [x] Any possible limitations, side effects, etc
- [x] Reference any other pull requests that might be related.

### CODE REVIEW

**Comment by mmoayyed** [comment]
Thanks for the changeset. I strongly recommend you reduce this pull request down to a laser-focused batch of changes targetting a specific functionality/module. Changing 80+ files at once is perfectly good, but chances of someone reviewing it, and taking the time to review it are very slim, which means your progress here will only be that much slower. This PR will sit here for a very long time.

**Comment by tduehr** [comment]
The problem I'm running into that may require a fairly large change is the JPA code all assumes hibernate backed by JDBC. That's improved somewhat since I switched from kundera to hibernate-ogm but may still require a few larger changes.

File: api/cas-server-core-api-configuration/src/main/java/org/apereo/cas/configuration/model/support/jpa/AbstractJpaProperties.java [hunk 1]
```diff
@@ -308,4 +339,24 @@ public long getFailFastTimeout() {
...
+
+    public static enum JpaType {
+        /**
+         * RDBMS type using hibernate.
+         */
+        JDBC,
+        /**
+         * NoSQL backends using Kundera.
+         */
+        CASSANDRA
```
Comment (mmoayyed) [inline api/cas-server-core-api-configuration/src/main/java/org/apereo/cas/configuration/model/support/jpa/AbstractJpaProperties.java]: This is likely not a good name for this behavior. Change to NOSQL or something like that.

File: api/cas-server-core-api-configuration/src/main/java/org/apereo/cas/support/jpa/DefaultJpaEntityManagerFactoryBeanFactory.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+    private DefaultJpaEntityManagerFactoryBeanFactory() {
+        jpaTypeMap = new ConcurrentHashMap<>();
+    }
+
+    public static DefaultJpaEntityManagerFactoryBeanFactory getInstance() {
+        return INSTANCE;
+    }
+
+    public void registerJEMFBF(AbstractJpaProperties.JpaType jpaType, JpaEntityManagerFactoryBeanFactory factory) {
```
[the remark this answers is not in the document]
Comment (mmoayyed) [reply api/cas-server-core-api-configuration/src/main/java/org/apereo/cas/support/jpa/DefaultJpaEntityManagerFactoryBeanFactory.java]: `registerFactory` would be better for the name.

File: api/cas-server-core-api-ticket/src/main/java/org/apereo/cas/ticket/TicketDefinition.java [hunk 3]
```diff
@@ -15,6 +15,14 @@
  * @since 5.1.0
  */
 public interface TicketDefinition extends Ordered {
+
+    /**
+     * Gets entity name.
+     *
+     * @return the entity name
+     */
+    String getTicketEntityName();
```
Comment (mmoayyed) [inline api/cas-server-core-api-ticket/src/main/java/org/apereo/cas/ticket/TicketDefinition.java]: I don't think this is actually needed at the API level; it's something very specific to a module and I believe the JDBC ticket registry and others do this by stuffing things into the Properties object. You may want to look into doing the same.

File: support/cas-server-support-jpa-core/src/main/java/org/apereo/cas/config/JpaEntityManagerFactoryBeanConfiguration.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+    @ConditionalOnMissingBean(name = "jpaEntityManagerFactoryBeanFactory")
+    @Autowired
+    @Bean
+    public DefaultJpaEntityManagerFactoryBeanFactory jpaEntityManagerFactoryBeanFactory(List<JpaEntityManagerFactoryBeanFactoryConfigurer> configurers) {
+        DefaultJpaEntityManagerFactoryBeanFactory factory = DefaultJpaEntityManagerFactoryBeanFactory.getInstance();
+
+        configurers.forEach(c -> {
+            final String name = StringUtils.removePattern(c.getClass().getSimpleName(), "\\$.+");
+            LOGGER.debug("Configuring logout execution plan [{}]", name);
```
Comment (mmoayyed) [inline support/cas-server-support-jpa-core/src/main/java/org/apereo/cas/config/JpaEntityManagerFactoryBeanConfiguration.java]: Wrong message?

File: api/cas-server-core-api-ticket/src/main/java/org/apereo/cas/ticket/TicketDefinition.java (hunk 3 above)
Comment (tduehr) [reply api/cas-server-core-api-ticket/src/main/java/org/apereo/cas/ticket/TicketDefinition.java]: Made more sense to me here as it's a property of the ticket and was pulled using a static method previously.

**Comment by tduehr** [comment]
Hibernate-OGM does not fully support abstracted CRUD.

**Pull request closed without merging**